xds/src/main/java/io/grpc/xds/ServerXdsClient.java

```diff
@@ -107,7 +107,7 @@ public void run() {
   @Nullable
   @Override
   Collection<String> getSubscribedResources(ResourceType type) {
-    if (newServerApi) {
+    if (newServerApi && (type == ResourceType.LDS)) {
```

Comment:
This is still incorrect. It should always return `null` for `type != ResourceType.LDS`.

Reply:
I can make the change but just curious what's the difference between returning null vs empty collection?

Reply:
It's not the same. I remember we discussed this before (in the doc).

In the old API, `ServerXdsClient` sends LDS requests with an empty list of resource names. When the RPC stream is recreated, we need to know if we need to re-subscribe to each type of resources. This is done by calling this `getSubscribedResources()`. It uses null vs non-null values to distinguish if the resources of that type is subscribed, and then decide if such a request needs to be sent. We cannot use empty vs non-empty to distinguish, because `ServerXdsClient` uses wildcard (empty list of resource names) to subscribe to LDS resources. 

Since `ServerXdsClient` never subscribes to resources other than LDS, it should always return `null` for other type of resources so that the abstract layer knows not to send requests for non-LDS resources.


The logic here should be

```java
if (type != ResourceType.LDS) {
  return null;
}
if (newServerApi) {
  return ImmutableList.<String>of(grpcServerResourceId);
}
return return Collections.emptyList();
```

Reply:
Strictly speaking, you should also include the condition of if there is a watcher watching for the resource. Just as what it initially looked like:
https://github.com/grpc/grpc-java/blob/40191b2f812e9173d6a967c360325ec90e62d1a0/xds/src/main/java/io/grpc/xds/ServerXdsClient.java#L90

Reply:
> Strictly speaking, you should also include the condition of if there is a watcher watching for the resource. Just as what it initially looked like:

The `ServerSdsClient` is only created when there is a watcher to be registered. I can't think of a case where ServerSdsClient exists and there is no watcher. So adding this check is not going to make a difference.